File: system-http-client/src/main/java/com/levnovikov/core_network/api_provider/ApiProvider.java
package com.levnovikov.core_network.api_provider;

import com.google.common.annotations.VisibleForTesting;
import com.levnovikov.core_network.HttpClient;
import com.levnovikov.core_network.request.Request;
import com.levnovikov.core_network.request.RequestBody;
import com.levnovikov.core_network.response.Response;

import java.io.IOException;
import java.util.Map;

import io.reactivex.Single;
import io.reactivex.annotations.Nullable;

/**
 * Created by lev.novikov
 * Date: 23/2/18.
 */

public class ApiProvider {

    private String baseUrl;
    private EntityConverter converter;
    private HttpClient client;
    private String contentType;

    private ApiProvider() {
    }

    public <Rq, Rs> Single<Rs> makeRequest(
            Request.Method method,
            String path, Rq requestEntity, Class<Rs> responseClass) {
        return client.makeCall(getRequest(method, path, requestEntity, null))
                .map(rsp -> getResponse(rsp, responseClass));

    }

    private <T> Request getRequest(Request.Method method, String path, T request, @Nullable Map<String, String> pathParams) {
        RequestBody body = new RequestBody(converter.convertTo(request), contentType);
        return new Request.Builder()
                .setMethod(method)
                .setPathParams(pathParams)
                .setUrl(baseUrl + path)
                .setBody(body)
                .build();
    }

    private <R> R getResponse(Response rsp, Class<R> responseClass) throws IOException {
        String body = rsp.getBody().getContentString();
        return converter.convertFrom(body, responseClass);
    }

    public static class Builder {
        private final ApiProvider provider = new ApiProvider();

        public Builder baseUrl(String baseUrl) {
            provider.baseUrl = baseUrl;
            return this;
        }

        public Builder converter(EntityConverter converter) {
            provider.converter = converter;
            return this;
        }

        public Builder client(HttpClient client) {
            provider.client = client;
            return this;
        }

        public Builder contentType(String contentType) {
            provider.contentType = contentType;
            return this;
        }

        public ApiProvider build() {
            return provider;
        }
    }
}

File: system-http-client/src/main/java/com/levnovikov/core_network/api_provider/ApiProvider.kt
package com.levnovikov.core_network.api_provider

import com.levnovikov.core_network.HttpClient
import com.levnovikov.core_network.exceptions.RequestError
import com.levnovikov.core_network.request.Request
import com.levnovikov.core_network.request.RequestBody
import com.levnovikov.core_network.response.Response
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException
import kotlin.reflect.KClass

/**
 * Created by lev.novikov
 * Date: 23/2/18.
 */

class ApiProvider private constructor() {

    private lateinit var baseUrl: String
    private lateinit var converter: EntityConverter
    private lateinit var errorConverter: ErrorConverter
    private lateinit var client: HttpClient
    private lateinit var contentType: String

    @Throws(RequestError::class)
    fun <Rq : Any, Rs : Any> makeRequest(
            method: Request.Method,
            path: String, requestEntity: Rq?, pathParams: Map<String, String>?, responseClass: KClass<Rs>): Rs {
        return getResponse(
                client.makeCall(
                        getRequest(method, path, requestEntity, pathParams)), responseClass)

    }

    private fun <T : Any> getRequest(method: Request.Method, path: String, request: T?, pathParams: Map<String, String>?): Request {
        return Request.Builder()
                .setMethod(method)
                .setPathParams(pathParams)
                .setUrl(baseUrl + path)
                .setBody(if (request != null) RequestBody(converter.convertTo(request), contentType) else null)
                .build()
    }

    @Throws(IOException::class, RequestError::class)
    private fun <R : Any> getResponse(rsp: Response, responseClass: KClass<R>): R {
        val body = rsp.body.contentString
        try {
            return converter.convertFrom(body, responseClass)
        } catch (e: JSONException) {
            throw RequestError(errorConverter.convertFrom(body))
        }
    }

    class Builder {
        private val provider = ApiProvider()

        fun baseUrl(baseUrl: String): Builder {
            provider.baseUrl = baseUrl
            return this
        }

        fun converter(converter: EntityConverter): Builder {
            provider.converter = converter
            return this
        }

        fun errorConverter(converter: ErrorConverter): Builder {
            provider.errorConverter = converter
            return this
        }

        fun client(client: HttpClient): Builder {
            provider.client = client
            return this
        }

        fun contentType(contentType: String): Builder {
            provider.contentType = contentType
            return this
        }

        fun build(): ApiProvider {
            return provider
        }
    }
}
